### ISSUE TITLE
Exception ignored in _del__

### ISSUE THREAD

**Issue by AndreaHasani** [zeromq/pyzmq#1424]
```
Exception ignored in: <function Socket.__del__ at 0x7f326bffce50>
Traceback (most recent call last):
  File "/home/strixx/.virtualenvs/telegram/lib/python3.8/site-packages/zmq/sugar/socket.py", line 67, in __del__
    self.close()
  File "/home/strixx/.virtualenvs/telegram/lib/python3.8/site-packages/zmq/_future.py", line 153, in close
    for event in list(chain(self._recv_futures, self._send_futures)):
TypeError: 'NoneType' object is not iterable
```

I've been getting this error for a while now. Cannot replicate it since it happens from time to time.
My version is latest commit (just built today since I saw a merge request for __close__)

**Comment by oddjobz** [comment]
Hi @AndreaHasani , I just noticed this, I've filed what looks to be the same issue in #1430 .. in my case I *think* it's caused in my case by trying to close a socket while the open is still in progress, so the "futures" variables are still assigned to "None" as per the class init code, rather then the iterators they become following an open.

**Comment by AndreaHasani** [comment]
@oddjobz In my case its easy it occurs from time to time when I restart my application I have a small chance to hit this. Basically it gives me that error and I must restart the app.

**Comment by oddjobz** [comment]
@AndreaHasani interesting .. I get this sometimes when processing a change, typically this means closing and reopening the socket very quickly. I'm just wondering whether there is a 'address in use' error lurking somewhere in the background ... I'll go back and see if I can trace it, mine happens fairly frequently, maybe 1 in 3 ...

**Comment by AndreaHasani** [comment]
@oddjobz Did you find anything about this?

**Comment by oddjobz** [comment]
Yes, I have something that seems to work. The issue seems to be (for me) that an exception is raised (for other reasons) while the open is underway, so when the exception causes the code to exit, it tries to close what is a "half open" socket. The solution (for me) is not to allow exception that terminate the process, so essentially instead of exiting on "KeyboardInterrupt", I trap the signal independently with the python signal package and set "self._running" to False .. and instead of "while True:" I now have "while self._running:" .. if that makes sense. So I now exit on the next iteration of the "while" rather than as a direct result of the exception being raised. Not had problems since then .. I guess it's an issue for anyone who relies on the signal to break a blocking system call, but I used timeouts for everything, so it's acceptable for me ...

**Comment by AndreaHasani** [comment]
I'm planning to switch to nng and check that out, this issue keeps crashing a process manager i use for managing over 800 processes and the only fix for when this happens is to restart. The problem is also that i can't isloate this problem properly locally

**Comment by oddjobz** [comment]
Erm, yeah, I looked at nng .. but I had issues with documentation and examples, and indeed confidence with regards to level of adoption and support. This bug is in "close", so another easy fix is to wrap all your "close" calls in a try / accept. That said, the 'real' problem might be something else entirely. In my code, I "am" trying to close the socket mid-way through an open, but that's a particular 'feature' of my application. If your code is trying to close a half-open connection, you might consider "why" it's trying to do that, as it would maybe imply an issue elsewhere linked to "why" the connection wasn't actually open .. ??

**Comment by AndreaHasani** [comment]
In mine I'm not even trying to close, it just randomly crashes. It baffles me that such a critical bug exist and no maintainer is looking into it since this would cause alot of issue. And its indeed a confirmed issue looking through zeromq pulls. 

Wrost case I'll will go with rest api solution since i do not need a huge throughput or lentancy

**Comment by oddjobz** [comment]
Yeah, but that was kind of my point. You're not trying to close, but it's crashing "in" close. This would imply (to me) that your connection open failed in some unexpected way and clean-up code (/the GC) is then calling close for you .. which would imply this bug is just a side-effect of the "real" issue, i.e. that the connection open is failing.

**Comment by AndreaHasani** [comment]
How do you wrap your calls of in a try..except? I mean I do not mind having to reload the socket or connection my only issue is that I have no idea on how to catch the issue when it happens

**Comment by minrk** [comment]
It would be super helpful to have a test case. Thanks for trying to figure it out. I'm a bit baffled as to what's unique about your use cases that causes this to come up, since it does seem to be very rare. I can't reproduce it anywhere, and don't really understand how it could happen, but #1444 should prevent the given exception, at least. As @oddjobz suggested, I do think this is only a side effect of the "real" issue, which is completely unknown without any context. Somehow `__init__` is not completing (or the object is partially disassembled, e.g. during process teardown), but `__del__` is being called instead of raising the exception in `__init__`. That's very weird! Adding errors to `__init__` does not appear to be enough to provoke it.

**Comment by AndreaHasani** [comment]
It would be hard to post all the setup here but what I'm doing is having a while loop over a asyncio function.
Here's the code in question which fails sometime (and I do not even get an exception)

```
    async def heartbeat(self):
        await asyncio.sleep(300)

        zeromq_server = self.ctx.socket(zmq.PAIR)
        zeromq_server.bind("ipc:///tmp/processManager_alive")
        zeromq_server.RCVTIMEO = 60000
        time.sleep(0.3)

        while True:
            try:
                await zeromq_server.recv()
            except zmq.error.Again:
                logger.critical(
                    "Heartbeat not received: Process Manager might be down."
                )
            except Exception:
                raise

            else:
                await zeromq_server.send(b"true")
```

**Comment by oddjobz** [comment]
Ok, I think I initially used something like this to squash the error message;
```python
        try:
          zeromq_server = self.ctx.socket(zmq.PAIR)
          zeromq_server.bind("ipc:///tmp/processManager_alive")
          zeromq_server.RCVTIMEO = 60000
          time.sleep(0.3)

          while True:
              try:
                  await zeromq_server.recv()
              except zmq.error.Again:
                  logger.critical(
                      "Heartbeat not received: Process Manager might be down."
                  )
              except Exception:
                  raise
              else:
                  await zeromq_server.send(b"true")
        finally:
          try:
            zeromq_server.close()
          except Exception:
            pass
```
As it stands, I would expect the "raise" to be your exit point, in which case if you look back at the stack trace, I would be hoping to see this close error as the "second" error reported, and maybe the "actual" error being listed prior .. i.e. raise quits, then the gc errors for a second time when it tries to close the socket.

**Comment by AndreaHasani** [comment]
@oddjobz Right I have not though about this, I though that it should never leave the While loop but your example makes sense.

**Comment by minrk** [comment]
I merged #1444 which should fix this. It would be great to have confirmation if it seems to fix the issue for you folks. Feel free to reopen if not.

**Comment by AndreaHasani** [comment]
I will pull the latest master on my app and see if this fixes it.
